### PR TITLE
Add support for oidc-discovery-provider ingress

### LINKED ISSUE #56: Ingress support for oidc-discovery
Need a templateable object for the subchart. It must be possible to expose the service outside the cluster.

### PR DESCRIPTION
**Pull request by kfox1111**
This patch enables exposing the oidc server out with an ingress along with tests to ensure it works.

Resolves #56

### CODE REVIEW

**Review by marcofranssen — CHANGES_REQUESTED** [review]
Couple of suggestions. will later today test it as well in my cluster to get a better feel on the UX of these new configurables.

File: charts/spire/charts/spiffe-oidc-discovery-provider/values.yaml [hunk 1]
```diff
@@ -134,3 +134,21 @@ telemetry:
         # limits:
         #   cpu: 100m
         #   memory: 64Mi
+
+ingress:
+  enabled: false
+  className: ""
+  annotations: {}
```
[the remark this answers is not in the document]
Comment (kfox1111) [reply charts/spire/charts/spiffe-oidc-discovery-provider/values.yaml:141]: its customary not to put in any annotations especially ones specific to any one ingress controller.

This config is straight out of helm create

File: charts/spire/charts/spiffe-oidc-discovery-provider/templates/ingress.yaml [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    {{- end }}
+  {{- end }}
+  rules:
+    {{- range .Values.ingress.hosts }}
+    - host: {{ .host | quote }}
+      http:
+        paths:
+          {{- range .paths }}
+          - path: {{ .path }}
+            {{- if and .pathType (semverCompare ">=1.18-0" $.Capabilities.KubeVersion.GitVersion) }}
```
[the remark this answers is not in the document]
Comment (kfox1111) [reply charts/spire/charts/spiffe-oidc-discovery-provider/templates/ingress.yaml:46]: Ok. I can pull that out. It was generated by helm create  standard. But, yeah. if we're requiring at least 1.21, I can remove it.

File: charts/spire/charts/spiffe-oidc-discovery-provider/values.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (dennisgove) [reply charts/spire/charts/spiffe-oidc-discovery-provider/values.yaml:141]: I agree about avoiding things specific to one ingress controller. Perhaps we could include in the example docs how one can set those annotations. That way we don't default any specific set in but do make it slightly easier for someone looking to set them.

[the remark this answers is not in the document]
Comment (kfox1111) [reply charts/spire/charts/spiffe-oidc-discovery-provider/values.yaml:141]: +1. Where should that go?

[the remark this answers is not in the document]
Comment (marcofranssen) [reply charts/spire/charts/spiffe-oidc-discovery-provider/values.yaml:141]: Looking at a security best practice considering we had huge discussions in the past on the spire-oidc component it always has to run on https I think we can't avoid this. @evan2645 @azdagron

File: charts/spire/charts/spiffe-oidc-discovery-provider/templates/ingress.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (marcofranssen) [reply charts/spire/charts/spiffe-oidc-discovery-provider/templates/ingress.yaml]: https://github.com/spiffe/helm-charts#dependencies-and-version-compatibility

we even only officialy support last 3 versions, basically same as Amazon does for EKS.

Unofficially we agreed in an earlier meeting if it doesn't cause additional maintenance we can support further back. In this case it does take additional logic for older versions, so I think we can remove that here.

File: charts/spire/charts/spiffe-oidc-discovery-provider/values.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (kfox1111) [reply charts/spire/charts/spiffe-oidc-discovery-provider/values.yaml:141]: I believe tls redirect is default in ingress-nginx? At least when you have a tls cert specified in the ingress.

There is no standard for forcing ssl in ingress though. so would have to find all the ways the various ingress controllers do it.

Also, there is no sensitive material in the oidc discovery agent. its just the public keys and endpoints.

File: charts/spire/charts/spiffe-oidc-discovery-provider/templates/ingress.yaml [hunk 3]
```diff
@@ -0,0 +1,41 @@
+{{- if .Values.ingress.enabled -}}
+{{- $fullName := include "spiffe-oidc-discovery-provider.fullname" . }}
+{{- $port := .Values.service.port }}
+apiVersion: networking.k8s.io/v1
+kind: Ingress
+metadata:
+  name: {{ $fullName }}
```
Comment (kfox1111) [inline charts/spire/charts/spiffe-oidc-discovery-provider/templates/ingress.yaml:7]: now that namespaceOverride has merged, we need one here.

**Review by marcofranssen — APPROVED** [review]
LGTM :rocket:

**Pull request merged**